### ISSUE TITLE
Add Mask  

### ISSUE THREAD

**Issue by li1u** [hbb1/2d-gaussian-splatting#52]
Hello, I'm sorry to bother you. How do you remove the background from the image? I inputted the corresponding mask image, but the result he rendered still includes the background

**Comment by li1u** [comment]
To be precise, I want to remove the background information from the reconstructed  3D model

**Comment by shenggedeqiang** [comment]
same question

**Comment by li1u** [comment]
其实关键在于作者裁剪模型的时候不太对，你可以看一下gaussian_surfels的代码，把他的裁剪模型的部分弄到这个里面，我是成功了，效果挺好的

**Comment by li1u** [comment]
然后读取掩模数据的时候也需要注意下

**Comment by hbb1** [comment]
I pack rgb with mask to be png images. The program will then treat the 4th channel as mask used for bounded TSDF fusion.

**Comment by ch1998** [comment]
> I pack rgb with mask to be png images. The program will then treat the 4th channel as mask used for bounded TSDF fusion.

In this way, can I directly use the RBGA image with the background removed to train?

**Comment by AndreaCastiella** [comment]
> > I pack rgb with mask to be png images. The program will then treat the 4th channel as mask used for bounded TSDF fusion.
> 
> In this way, can I directly use the RBGA image with the background removed to train?

Hello! Did you manage to train without background using rgba images?

**Comment by hbb1** [comment]
I think it is trival to do so. Simply packing the rendered_rgb and render_alpha we get an RGBA image. Next, supervise it with the GT RGBA image.

**Comment by Zerui-Yu** [comment]
> 其实关键在于作者裁剪模型的时候不太对，你可以看一下gaussian_surfels的代码，把他的裁剪模型的部分弄到这个里面，我是成功了，效果挺好的

请问你是指这些代码吗？把gaussian_surfels的裁剪策略移植到这里？
2d-gaussian-splatting，train.py(line 126-135)

            if iteration < opt.densify_until_iter:
                gaussians.max_radii2D[visibility_filter] = torch.max(gaussians.max_radii2D[visibility_filter], radii[visibility_filter])
                gaussians.add_densification_stats(viewspace_point_tensor, visibility_filter)

                if iteration > opt.densify_from_iter and iteration % opt.densification_interval == 0:
                    size_threshold = 20 if iteration > opt.opacity_reset_interval else None
                    gaussians.densify_and_prune(opt.densify_grad_threshold, opt.opacity_cull, scene.cameras_extent, size_threshold)
                
                if iteration % opt.opacity_reset_interval == 0 or (dataset.white_background and iteration == opt.densify_from_iter):
                    gaussians.reset_opacity()

[gaussian_surfels](https://github.com/turandai/gaussian_surfels)，train.py(line 174-184)


            if iteration > opt.densify_from_iter:
                # Keep track of max radii in image-space for pruning
                gaussians.max_radii2D[visibility_filter] = torch.max(gaussians.max_radii2D[visibility_filter], radii[visibility_filter])
                gaussians.add_densification_stats(viewspace_point_tensor, visibility_filter)
                min_opac = 0.1
                if iteration % opt.densification_interval == 0:
                    gaussians.adaptive_prune(min_opac, scene.cameras_extent)
                    gaussians.adaptive_densify(opt.densify_grad_threshold, scene.cameras_extent)
                
                if (iteration - 1) % opt.opacity_reset_interval == 0 and opt.opacity_lr > 0:
                    gaussians.reset_opacity(0.12, iteration)

**Comment by sidoniepc** [comment]
> I think it is trival to do so. Simply packing the rendered_rgb and render_alpha we get an RGBA image. Next, supervise it with the GT RGBA image.

Hello, do you mean that we don't need to modify anything in the code, just use RGBA images as input?

**Comment by oUp2Uo** [comment]
Hi,
I am using RGBA png files as input,
and I have tried the origin code, and edited https://github.com/hbb1/2d-gaussian-splatting/blob/df1f6c684cc4e41a34937fd45a7847260e9c6cd7/scene/cameras.py#L43C1-L48C38, 
both have the same result,
which still have background.
Has any one managed to run with mask?
Thank you.

**Comment by oUp2Uo** [comment]
> 其实关键在于作者裁剪模型的时候不太对，你可以看一下gaussian_surfels的代码，把他的裁剪模型的部分弄到这个里面，我是成功了，效果挺好的

@li1u 请问有详细的做法么？
目前2dgs的这个做法是把背景强行变成黑的，然后跑完了之后会有很多黑色的大块，和想象中的mask效果不太一样。
gaussian_surfels是真正的扣掉了背景部分么？我试读了下代码发现和2dgs的不太一样，好像mask他有自己单独处理的很多个地方

**Comment by AndreaCastiella** [comment]
I made some changes to the code. I added the alpha mask in the fourth channel and then modified the code so that it packs the rendered alpha into the rendered rgb image and also modified the GT RGBA image so that the parts where alpha is 0 the image is the same colour as the one chosen for the backgorund. Here is how I modified the code. Hope it helps :)

Expand background:
`   ... 
    bg_color = [1, 1, 1] if dataset.white_background else [0, 0, 0]
    background = torch.tensor(bg_color, dtype=torch.float32, device="cuda")

    # Expand background to match the shape of the image
    background_expanded = background.unsqueeze(1).unsqueeze(2)
    # Assuming the first camera to initialize dimensions
    viewpoint_cam = scene.getTrainCameras()[0]
    sample_image = viewpoint_cam.original_image.cuda()
    background_expanded = background_expanded.expand_as(sample_image)

    iter_start = torch.cuda.Event(enable_timing=True)
    iter_end = torch.cuda.Event(enable_timing=True)
   ...
`

And add the fourth channel:

`        ...
        image, viewspace_point_tensor, visibility_filter, radii, alpha, depth = (
            render_pkg["render"],
            render_pkg["viewspace_points"],
            render_pkg["visibility_filter"],
            render_pkg["radii"],
            render_pkg["rend_alpha"],
            render_pkg["surf_depth"]
        )
        # Get the ground truth image and alpha mask
        gt_image = viewpoint_cam.original_image.cuda()
        gt_alpha = (
            viewpoint_cam.gt_alpha_mask.cuda()
            if viewpoint_cam.gt_alpha_mask is not None
            else None
        )
        # Set the pixels in the GT image where alpha is 0 to background color
        if gt_alpha is not None:
            gt_image = torch.where(gt_alpha == 0, background_expanded, gt_image) # Background will be set to bg color
            # Concatenate alpha to the RGB images
            image = torch.cat([image, alpha], dim=0)
            gt_image = torch.cat([gt_image, gt_alpha], dim=0)

        Ll1 = l1_loss(image, gt_image)
        loss = (1.0 - opt.lambda_dssim) * Ll1 + opt.lambda_dssim * (
            1.0 - ssim(image, gt_image)
        )
        ...
`

**Comment by wslgqq277g** [comment]
> I made some changes to the code. I added the alpha mask in the fourth channel and then modified the code so that it packs the rendered alpha into the rendered rgb image and also modified the GT RGBA image so that the parts where alpha is 0 the image is the same colour as the one chosen for the backgorund. Here is how I modified the code. Hope it helps :)
> 
> Expand background:
> 
> ```
>     bg_color = [1, 1, 1] if dataset.white_background else [0, 0, 0]
>     background = torch.tensor(bg_color, dtype=torch.float32, device="cuda")
> 
>     # Expand background to match the shape of the image
>     background_expanded = background.unsqueeze(1).unsqueeze(2)
>     # Assuming the first camera to initialize dimensions
>     viewpoint_cam = scene.getTrainCameras()[0]
>     sample_image = viewpoint_cam.original_image.cuda()
>     background_expanded = background_expanded.expand_as(sample_image)
> 
>     iter_start = torch.cuda.Event(enable_timing=True)
>     iter_end = torch.cuda.Event(enable_timing=True)
>    ...
> ```
> 
> And add the fourth channel:
> 
> ```
>         image, viewspace_point_tensor, visibility_filter, radii, alpha, depth = (
>             render_pkg["render"],
>             render_pkg["viewspace_points"],
>             render_pkg["visibility_filter"],
>             render_pkg["radii"],
>             render_pkg["rend_alpha"],
>             render_pkg["surf_depth"]
>         )
>         # Get the ground truth image and alpha mask
>         gt_image = viewpoint_cam.original_image.cuda()
>         gt_alpha = (
>             viewpoint_cam.gt_alpha_mask.cuda()
>             if viewpoint_cam.gt_alpha_mask is not None
>             else None
>         )
>         # Set the pixels in the GT image where alpha is 0 to background color
>         if gt_alpha is not None:
>             gt_image = torch.where(gt_alpha == 0, background_expanded, gt_image) # Background will be set to bg color
>             # Concatenate alpha to the RGB images
>             image = torch.cat([image, alpha], dim=0)
>             gt_image = torch.cat([gt_image, gt_alpha], dim=0)
> 
>         Ll1 = l1_loss(image, gt_image)
>         loss = (1.0 - opt.lambda_dssim) * Ll1 + opt.lambda_dssim * (
>             1.0 - ssim(image, gt_image)
>         )
>         ...
> ```

Is the outcome result good?  Thanku for posting your modified code, I'd like to try it on my own dataset later.

**Comment by malfonsoarquimea** [comment]
Yes, the result is quite good :) I hope you can make it work!

**Comment by AndreaCastiella** [comment]
Yes, the result is quite good :) I hope you can make it work!

**Comment by Tiandishihua** [comment]
> 其实关键在于作者裁剪模型的时候不太对，你可以看一下gaussian_surfels的代码，把他的裁剪模型的部分弄到这个里面，我是成功了，效果挺好的

大佬，您好！请问你还是读取colmap格式数据，然后添加的mask功能吗？能加V沟通一下吗
